rfcs/0074-tvmscript-unified-printer.md

```diff
@@ -0,0 +1,487 @@
+
+**Goal 2 [Third-Party IRs in Multi-Stage Compilation].** Modularize and
+infrastructuralize the printer to support more future IRs or third-party IRs at
+any level with maintainability, for example, IRs at lower-level than TIR, or
+Relax VM executable.
+
+The current TVMScript printer is tightly coupled with TIR by being a subclass
+of TIR-specific functors. This design isn’t scalable when we want to support
+more IRs
+
+```c++
+class TVMScriptPrinter : public tir::StmtFunctor<Doc(const tir::Stmt&)>,
+                         public tir::ExprFunctor<Doc(const PrimExpr&, ExprPrecedence*)>,
+                         public TypeFunctor<Doc(const Type&)>  
+{...}
+```
+
+More importantly, it’s impossible for the current approach to support
+third-party IR bteing registered in a dynamic library.
+
+**Goal 3 [Reproducibility and Error Reporting].** Expand reproducibility and
+flexible rendering of diagnostic messages during any level of IR
+transformation.
+
+For example, the following snippet runs and produces an error.
+
+```py
+import tvm
+
+@T.prim_func
+def func_a(A: T.Buffer[(1,), "int32"]):
+    A[0] = 0
+
+@T.prim_func
+def func_b(A: T.Buffer[(8,), "int32"]):
+    A[0] = 0
+
+tvm.ir.assert_structural_equal(func_a, func_b)
+```
+
+The current error message indicates what the difference was, but not where it
+occurred.  This can sometimes be inferred from a stack trace, but becomes
+increasingly difficult with larger IR graphs.
+
+```
+ValueError: StructuralEqual check failed, caused by lhs:
+1
+and rhs:
+8
+```
+
+TUNIP should enable individual utilities and IR passes to have error messages
+directing the user to exact locations in the IR representation.
+
+```
+ValueError: StructuralEqual check failed, first delta highlighted below
+
+@T.prim_func
+def func_a(A: T.Buffer[(1,), "int32"]) -> None:
+                       ^^^^
+    A[0] = 0
+
+@T.prim_func
+def func_b(A: T.Buffer[(8,), "int32"]) -> None:
+                       ^^^^
+    A[0] = 0
+```
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This section introduces the design philosophy of the printer, and demonstrates
+the proposed user-facing APIs where users means IR developers.
+
+## Two-Stage Translation
+
+Traditionally in TVM stack, printing is a single-stage process. The printer
+assumes certain syntax of the target language, and therefore, so far there are
+3 different printers all for TIR: ReprPrinter, TIRTextPrinter,
+TVMScriptPrinter.
+
+The Doc DSL (not to be confused with the Doc class at
+[https://github.com/apache/tvm/blob/main/src/printer/doc.h#L67](https://github.com/apache/tvm/blob/main/src/printer/doc.h#L67))
+is introduced to allow better consistency and scalability. An IR, which could
+be TIR, Relax or any other ones developed by third-party vendors, is first
+translated to an intermediate Doc DSL, and then the DSL is mapped to a target
+language, for example, Python, C++ IRBuilder API, or Rust.
```

Comment:
Let's reword it a little bit to minimize the novelty (industry doesnt like novel things). Say something like we extend the idea of the existing Doc class

Reply:
Updated